### PR TITLE
[WIP] new melt formulation

### PR DESCRIPTION
**Pull request by jdannberg**
This is a first try to change our melt transport formulation to the one from https://doi.org/10.1029/2023JB027469
@tjhei Can you have a look? I used the melt_transport_convergence_test to test the implementation, but the results are quite wrong (the velocities a little bit, the pressures by a lot). For now, I also hard-coded e to be 1.e-5 / viscosity_c; in the paper they use 1.e-10 / viscosity_c as far as I can tell, but if I make e any smaller the solver does not converge any more. That is even though this problem never goes to small melt fractions, so we should not even need any regularization. And for the testcase, I thought that even if I made a mistake in the computation of the pressure somehow, at least the solid velocity should still be correct, which it isn't. 

EDIT: We now think the solution is correct, based on the fact that we get the correct velocities and fluid pressure for the melt_transport_convergence_simple test. The other pressures are off by a constant factor, probably due to the pressure normalization. 

Now the question is: Is this new formulation actually better than what we have?
One issue is that we cannot normalize the fluid pressure any more, because it explicitly shows up in the equations. In addition, the fluid pressure is defined to be zero in regions without melt, which makes it jump at the melt/no melt boundary. Therefore, it might be better to apply the regularization to both p_f and p_t in the second equation. This would make the fluid pressure equal to the total pressure for zero porosity (no jump) and we can again normalize the pressure because only p_t - p_f appears in the equations, so we could change both by the same amount without changing the solution otherwise.

This still means that in the case of zero porosity, the second equation would be eps (p_t - p_f) = 0 with a small eps (in the paper, they choose 1e-10). We need to see if that is a problem for the linear solver. 

Other things we still need to do:
* [x] Try with regularization -- we need the additional regularization (whole p_f row), otherwise the fluid pressure values have a really big spike at the interface between melt and not melt regions 
* [x] Pressure normalization (apply to both total and fluid pressure: let p_t be zero at the surface and apply the normalization to p_f as well)
* [ ] change porosity and other advection equations
* [x] remove melt_cell
* [ ] rename solution variables (e.g. ex_p_c)
* [ ] Compute inverse compaction pressure rather than compaction pressure in the material models
* [ ] Update the cookbooks/tests accordingly: Some material models had minimum porosities for the compaction viscosity; these should now instead be translated into how large the regularization is. For reference: Reactive fluid transport: 1e-8. Katz 2003: 1e-4. Shear bands: 1e-4. Solubility: 1e-8. 
* [ ] Think about scaling for regularization (it should scale with some reference viscosity, but we cannot use the actual compaction viscosity because that becomes infinity at some point)
* [x] do not average divergence of velocity for melt advection
* [ ] think about averaging schemes for inverse compaction viscosity
* [ ] fix make_pressure_rhs_compatible (we probably have to do something different for p_t that what we did for p_c)

Test cases to do:
* [x] Convergence test for the Arbogast with both May and our new regularization
* [ ] Check if solid pressure is correct
* [x] Solitary wave
* [ ] Time-dependent global 2D box --> does not work because compaction viscosity does not go to infinity for zero porosity
* [x] Time-dependent mid-ocean ridge
* [ ] Update the analytical solutions for all p_c cases + recompute the RHS terms for the manufactured solutions for p_c

Already done:
* [ ] think about the preconditioner
* [x] Assemble new terms
* [x] remove K_D scaling (p_c_scale)
* [x] recompute other melt outputs: Melt velocity is already correct, solid pressure has been changed
* [x] coupling
* [x] we checked that the boundary conditions should be okay as they are

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/simulator/melt.cc [hunk 1]
```diff
@@ -299,12 +248,12 @@ namespace aspect
...
                 data.local_matrix(i,j) +=
                   (
-                    (p_c_scale * one_over_eta *
+                    (one_over_eta *
                      pressure_scaling *
                      pressure_scaling)
                     * scratch.phi_p[i] * scratch.phi_p_c[j]
                     +
-                    (p_c_scale * one_over_eta *
+                    (one_over_eta *
```
Comment (tjhei) [inline source/simulator/melt.cc]: we will need to rename scratch.phi_p_c ...

File: source/simulator/melt.cc [hunk 2]
```diff
@@ -487,9 +430,10 @@ namespace aspect
           const unsigned int porosity_index = introspection.compositional_index_for_name("porosity");
           const double porosity = std::max(scratch.material_model_inputs.composition[q][porosity_index],0.000);
 
-          const double K_D = this->get_melt_handler().limited_darcy_coefficient(melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q],
-                                                                                p_c_scale > 0);
+          const double K_D = melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q];
           const double viscosity_c = melt_outputs->compaction_viscosities[q];
+          const double e = 1.e-5 / viscosity_c;
+          const double regularization = 1./std::sqrt(1./(viscosity_c*viscosity_c)+e*e);
```
Comment (tjhei) [inline source/simulator/melt.cc]: Is the 1/ correct?

**Review by tjhei — APPROVED** [review]

File: source/simulator/melt.cc (hunk 2 above)
Comment (tjhei) [reply source/simulator/melt.cc]: can you multiply with ``regularization`` below and remove the outer 1/ ?

File: source/simulator/melt.cc [hunk 3]
```diff
@@ -289,7 +238,7 @@ namespace aspect
                                              scratch.grad_phi_p[i] *
```
Comment (jdannberg) [inline source/simulator/melt.cc:245]: Line 230: 
replace one_over_eta with one_over_viscosity_c_e

File: source/simulator/melt.cc [hunk 4]
```diff
@@ -289,7 +238,7 @@ namespace aspect
...
                                              scratch.grad_phi_p[j]
                                              +
-                                             (1./eta + 1./viscosity_c) * p_c_scale * p_c_scale *
+                                             (1./eta + 1./viscosity_c) *
```
Comment (jdannberg) [inline source/simulator/melt.cc]: Maybe 1./viscosity_c needs to be -1./viscosity_c
Edit: Nevermind, it should stay a plus

File: source/simulator/melt.cc [hunk 5]
```diff
@@ -299,12 +248,12 @@ namespace aspect
                 // add S between p_c and p_f
                 data.local_matrix(i,j) +=
                   (
-                    (p_c_scale * one_over_eta *
+                    (one_over_eta *
```
Comment (jdannberg) [inline source/simulator/melt.cc]: times (-1) and this should be one_over_viscosity_c

File: source/simulator/melt.cc (hunk 1 above)
Comment (jdannberg) [inline source/simulator/melt.cc]: times (-1) and this should be one_over_viscosity_c

File: source/simulator/melt.cc [hunk 6]
```diff
@@ -299,12 +249,12 @@ namespace aspect
                 // add S between p_c and p_f
                 data.local_matrix(i,j) +=
                   (
-                    (p_c_scale * one_over_eta *
+                    (-1./viscosity_c *
```
Comment (jdannberg) [inline source/simulator/melt.cc]: Add the e_squared here.

**Comment by jdannberg** [comment]
For testing this in realistic models

```
subsection Discretization
  subsection Stabilization parameters
    set Use artificial viscosity smoothing = true
    set beta = 0.052, 0.1, 0.052
    set cR = 0.11, 0.2, 0.11
  end
end
```
(the values that are different from the default, in the case above, the second in the list, are for the porosity)
and

```
subsection Melt settings
  set Compaction viscosity regularization factor = 1e-28
end
```
(this is the inverse of the compaction viscosity at zero porosity)

**Comment by jdannberg** [comment]
Status for today:
- Realistic problems that have zero porosity along top/bottom boundaries (kind of) work with additional stabilization (see above); also if one of the boundaries is open and porosity is nonzero along the boundaries
- Incompressible models with closed boundaries where the boundary conditions are compatible work (that means that for complicated boundary conditions, they only work in high resolution)
- Incompressible models with boundary conditions that are not compatible (not even analytically) do not give a reasonable solution

What still needs to be done:
- [ ] For compressible models, there is a term missing (in the last equation)
- [ ] Test solver performance and robustness with mid0ocean ridge example from 2019 paper (2D and 3D)
also see above. The file is called mid_ocean_ridge_ascii_data.prm and it has an open bottom so it should be compatible.

**Comment by bangerth** [comment]
So did the rhs correction help?
